File: 3433-count-mentions-per-user.java
import java.util.*;

class CountMentionsPerUser {

    public int[] countMentions(int numberOfUsers, List<List<String>> events) {
        events.sort((a, b) -> {
            if (Integer.parseInt(a.get(1)) == Integer.parseInt(b.get(1))) {
                return b.get(0).compareTo(a.get(0));
            }
            return Integer.parseInt(a.get(1)) - Integer.parseInt(b.get(1));
        });

        int[] mentions = new int[numberOfUsers];
        int ONLINE_TIME_OFFSET = 60;
        int[] usersOnlineTime = new int[numberOfUsers];
        int allMentioned = 0;

        for (List<String> event : events) {
            String type = event.get(0), users = event.get(2);
            int time = Integer.parseInt(event.get(1));
            if (type.equals("MESSAGE")) {
                if (users.equals("HERE")) {
                    for (int user = 0; user < numberOfUsers; user++) {
                        int onlineTime = usersOnlineTime[user];
                        if (time >= onlineTime) {
                            mentions[user]++;
                        }
                    }
                    continue;
                }
                if (users.equals("ALL")) {
                    allMentioned++;
                    continue;
                }
                String[] userList = users.split(" ");
                for (String user : userList) {
                    int userId = Integer.parseInt(user.substring(2));
                    mentions[userId]++;
                }
                continue;
            }
            usersOnlineTime[Integer.parseInt(users)] = time + ONLINE_TIME_OFFSET;
        }

        for (int user = 0; user < numberOfUsers; user++) {
            mentions[user] += allMentioned;
        }

        return mentions;
    }

    public static void main(String[] args) {
        CountMentionsPerUser sol = new CountMentionsPerUser();
        System.out.println("Output: " + Arrays.toString(sol.countMentions(2, Arrays.asList(Arrays.asList("MESSAGE","10","id1 id0"),Arrays.asList("OFFLINE","11","0"),Arrays.asList("MESSAGE","71","HERE")))) + ", expected: [2, 2]");
        System.out.println("Output: " + Arrays.toString(sol.countMentions(2, Arrays.asList(Arrays.asList("MESSAGE","10","id1 id0"),Arrays.asList("OFFLINE","11","0"),Arrays.asList("MESSAGE","12","ALL")))) + ", expected: [2, 2]");
        System.out.println("Output: " + Arrays.toString(sol.countMentions(2, Arrays.asList(Arrays.asList("OFFLINE","10","0"),Arrays.asList("MESSAGE","12","HERE")))) + ", expected: [0, 1]");
    }
}

File: 3433-count-mentions-per-user.go
//go:build ignore

package main

import (
	"cmp"
	"fmt"
	"slices"
	"strconv"
	"strings"
)

func countMentions(numberOfUsers int, events [][]string) []int {
	slices.SortFunc(events, func(e1, e2 []string) int {
		time1, _ := strconv.Atoi(e1[1])
		time2, _ := strconv.Atoi(e2[1])
		if time1 == time2 {
			return cmp.Compare(e2[0], e1[0])
		}
		return cmp.Compare(time1, time2)
	})

	mentions := make([]int, numberOfUsers)
	onlineTimeOffset := 60
	usersOnlineTime := make([]int, numberOfUsers)
	allMentioned := 0

	for _, event := range events {
		eventType, time, users := event[0], event[1], event[2]
		if eventType == "MESSAGE" {
			if users == "HERE" {
				for user, onlineTime := range usersOnlineTime {
					timeInt, _ := strconv.Atoi(time)
					if timeInt >= onlineTime {
						mentions[user]++
					}
				}
				continue
			}
			if users == "ALL" {
				allMentioned++
				continue
			}
			userList := strings.Split(users, " ")
			for _, user := range userList {
				userId, _ := strconv.Atoi(user[2:])
				mentions[userId]++
			}
			continue
		}
		userId, _ := strconv.Atoi(users)
		timeInt, _ := strconv.Atoi(time)
		usersOnlineTime[userId] = timeInt + onlineTimeOffset
	}
	for user := range numberOfUsers {
		mentions[user] += allMentioned
	}

	return mentions
}

func main() {
	fmt.Printf("Output: %v, expected: [2, 2]\n", countMentions(2, [][]string{{"MESSAGE", "10", "id1 id0"}, {"OFFLINE", "11", "0"}, {"MESSAGE", "71", "HERE"}}))
	fmt.Printf("Output: %v, expected: [2, 2]\n", countMentions(2, [][]string{{"MESSAGE", "10", "id1 id0"}, {"OFFLINE", "11", "0"}, {"MESSAGE", "12", "ALL"}}))
	fmt.Printf("Output: %v, expected: [2, 2]\n", countMentions(2, [][]string{{"OFFLINE", "10", "0"}, {"MESSAGE", "12", "HERE"}}))
}
